### ISSUE TITLE
16-bit quantization issue

### ISSUE THREAD

**Issue by lehahoang** [NervanaSystems/distiller#319]
My question is whether Distiller support 16-bit quantization or not?
It works perfectly well with 8-bit quantization but things go wrong with 16-bit quantization scheme.
As I check the parameters of the wrapped modules from "quantized_checkpoint.pth.tar" file, only some of them are "truly" integer numbers. 
My model is resnet20 and my command line is:
python3 compress_classifier.py --arch resnet20_cifar ../data.cifar --qebw 16 --qeba 16 --resume-from ./resnet20/resnet20_32fp_cifar.pth.tar --quantize-eval --evaluate --qe-mode sym

Thank you.

**Comment by guyjacob** [comment]
Hi @lehahoang,

You shouldn't be seeing any difference between 8 and 16-bits in terms of parameter values being integer numbers. But you need to make sure you're looking at the right parameters:
* You should look only at parameters that end with ".weight" and **".base_b_q"**. Note you shouldn't look at ".bias". In short, since you're running dynamic quantization (not passing a stats file via `--qe-stats-file`), we can only save a "baseline" quantized bias, and we re-quantize that baseline with each new batch that comes in. We save that baseline in the ".base_b_q" parameter. The math behind this is detailed [here](https://nervanasystems.github.io/distiller/algo_quantization.html#range-based-linear-quantization).
* Only convolution and linear layers are relevant. We don't quantize BatchNorm layers. So BN parameters stay in FP32. (Note that if you're running the latest code from master, which already has batch-norm folding, this shouldn't be an issue with ResNet-20. All BN modules should be eliminated prior to quantization)

I just tried and it works fine for me. Considering the points above, do you still see the problem?

**Comment by guyjacob** [comment]
Here's the code I used to check:
```python
import torch
import distiller

cp_8b = torch.load('PATH-TO-8B-QUANTIZED-CHECKPOINT)
cp_16b = torch.load('PATH-TO-16B-QUANTIZED-CHECKPOINT)

for k in cp_8b['state_dict']:
    if not(k.endswith('weight') or k.endswith('base_b_q')):
        continue
    p_8b = cp_8b['state_dict'][k]
    p_16b = cp_16b['state_dict'][k]
    print(k)
    print(' 8b: ', (p_8b.round() == p_8b).all().item())
    print(' 16b:', (p_16b.round() == p_16b).all().item())
```

You should see an output with all 1's, something like:
```
module.conv1.base_b_q
 8b:  1
 16b: 1
module.conv1.wrapped_module.weight
 8b:  1
 16b: 1
module.layer1.0.conv1.base_b_q
 8b:  1
 16b: 1
module.layer1.0.conv1.wrapped_module.weight
 8b:  1
 16b: 1
...
```

**Comment by lehahoang** [comment]
Hi Jacob,
Thanks for your kind reply. I have considered all the points you enlisted but problem is unsolved at the moment. I have followed the instruction from this[ link](https://github.com/NervanaSystems/distiller/blob/master/examples/quantization/post_train_quant/command_line.md):


To make it more clear, I would like to summarize steps I have done.

- **Step 1:** I used command lines below to generate 16-bit quantized model:

$ python3 compress_classifier.py --arch resnet20_cifar ../data.cifar –resume-from resnet20/resnet20_32fp_cifar.pth.tar  --qe-calibration 0.05

$ python3 compress_classifier.py --arch resnet20_cifar ../data.cifar –resume-from resnet20/resnet20_32fp_cifar.pth.tar  --evaluate --quantize-eval –qe-stats-file acts_quantization_stats.yaml --qe-per-channel --qe-bits-acts 16 --qe-bits-wts 16

Please note that resnet20_32fp_cifar.pth.tar is the model I trained by myself.

- **Step 2**: I took note the log file of quantization process 

Log file:
Loading activation stats from: acts_quantization_stats.yaml
Preparing model for quantization using PostTrainLinearQuantizer
Parameter 'module.conv1.wrapped_module.weight' will be quantized to 16 bits
Parameter 'module.layer1.0.conv1.wrapped_module.weight' will be quantized to 16 bits
Parameter 'module.layer1.0.conv2.wrapped_module.weight' will be quantized to 16 bits
Parameter 'module.layer1.1.conv1.wrapped_module.weight' will be quantized to 16 bits
Parameter 'module.layer1.1.conv2.wrapped_module.weight' will be quantized to 16 bits
Parameter 'module.layer1.2.conv1.wrapped_module.weight' will be quantized to 16 bits
Parameter 'module.layer1.2.conv2.wrapped_module.weight' will be quantized to 16 bits
Parameter 'module.layer2.0.conv1.wrapped_module.weight' will be quantized to 16 bits
Parameter 'module.layer2.0.conv2.wrapped_module.weight' will be quantized to 16 bits
Parameter 'module.layer2.0.downsample.0.wrapped_module.weight' will be quantized to 16 bits
Parameter 'module.layer2.1.conv1.wrapped_module.weight' will be quantized to 16 bits
Parameter 'module.layer2.1.conv2.wrapped_module.weight' will be quantized to 16 bits
Parameter 'module.layer2.2.conv1.wrapped_module.weight' will be quantized to 16 bits
Parameter 'module.layer2.2.conv2.wrapped_module.weight' will be quantized to 16 bits
Parameter 'module.layer3.0.conv1.wrapped_module.weight' will be quantized to 16 bits
Parameter 'module.layer3.0.conv2.wrapped_module.weight' will be quantized to 16 bits
Parameter 'module.layer3.0.downsample.0.wrapped_module.weight' will be quantized to 16 bits
Parameter 'module.layer3.1.conv1.wrapped_module.weight' will be quantized to 16 bits
Parameter 'module.layer3.1.conv2.wrapped_module.weight' will be quantized to 16 bits
Parameter 'module.layer3.2.conv1.wrapped_module.weight' will be quantized to 16 bits
Parameter 'module.layer3.2.conv2.wrapped_module.weight' will be quantized to 16 bits
Parameter 'module.fc.wrapped_module.weight' will be quantized to 16 bits
Parameter 'module.fc.wrapped_module.bias' will be quantized to 32 bits

- **Step 3:** I loaded the quantized model and checked the parameters in [‘state_dict’]

module.conv1.wrapped_module.weight
module.bn1.weight
module.layer1.0.conv1.wrapped_module.weight
module.layer1.0.bn1.weight
module.layer1.0.conv2.wrapped_module.weight
module.layer1.0.bn2.weight
……

- **Step 4:** I print out the values of wrapped model and received the result as follow

$ print(cp_16b['state_dict']['module.conv1.wrapped_module.weight'])

tensor([[[[-1.1674e+04,  1.3735e+04,  1.7130e+04],
          [-2.1960e+04, -2.3083e+04,  3.1799e+04],
          [-2.3258e+04, -2.0422e+04,  1.6858e+04]],
…….

Step 5: I tried your test code and here is the result:
module.conv1.wrapped_module.weight
 8b:  1
module.bn1.weight
 8b:  0
module.layer1.0.conv1.wrapped_module.weight
 8b:  1
module.layer1.0.bn1.weight
 8b:  0
…..


_It is a bit unusual that the parameter ends with ‘b_q’ is missing in 16-bit quantization scheme. However, I found it in case of 8-bit quantization scheme_

Do you think I have missed any important step?
Thanks

**Comment by guyjacob** [comment]
You didn't miss any steps, and actually everything you're seeing is as expected:
* _tensor([[[[-1.1674e+04, 1.3735e+04, 1.7130e+04],
...
module.conv1.wrapped_module.weight
8b: 1_
So actually we can see that the weights are indeed integer values. When I saw the tensor values that were printed I was initially confused, then realized these are actually integer values: -1.1674e+04 = 11674, etc. They're just printed in scientific notation.
* _module.bn1.weight
8b: 0_
As I mentioned above, batch norm weights are not quantized. So this is also expected. Please update to the latest version of the code - we added batch-norm folding last week. When you run with the latest code these BN parameters will be gone from the quantized checkpoint.
* Regarding `base_b_q` - The fact that you're not seeing it is not related to 16 vs. 8 bit. This is indeed confusing, I'll try to explain.
The original command line you posted was using "dynamic" quantization, and as I explained in my first response, one of the implications of that is that we save this `base_b_q` parameter, which we re-quantize for every batch.
This time you ran with "static" quantization - you used a stats file. In this case, we can do a one-time quantization of the bias upfront. We don't need to re-quantize for every new input. So in this case we don't save the `base_b_q` parameter. We directly quantize the `bias` parameter. So if you'd change the code I posted above to look for `bias` instead of `base_b_q`, you'll see those parameters have integer values as well. Again, I realize this is confusing - I hope I was able to clarify what's going on.

**Comment by lehahoang** [comment]
Hi Jacob,

Thanks for your clarification. I think this problem is solve now.
